src/frontend/src/flows/iframeWebAuthn.ts

```diff
@@ -0,0 +1,188 @@
+import { Connection } from "$src/utils/iiConnection";
+import {
+  waitForWindowReadyRequest,
+  waitForWindowReadyResponse,
+} from "$src/utils/internalPostMessage";
+import { isNullish } from "@dfinity/utils";
+
+export const WEBAUTHN_IFRAME_PATH = "#iframe/webauthn";
+
+interface CredentialRequest {
+  ii_credential_request: {
+    id: string;
+    options: CredentialRequestOptions;
+  };
+}
+
+interface CredentialResponse {
+  ii_credential_response: {
+    id: string;
+  } & (
+    | {
+        result: Omit<PublicKeyCredential, "getClientExtensionResults">;
+      }
+    | {
+        error: string;
+      }
+  );
+}
+
+const isCredentialRequest = (data: unknown): data is CredentialRequest =>
+  typeof data === "object" && data !== null && "ii_credential_request" in data;
+
+const isCredentialResponse = (data: unknown): data is CredentialResponse =>
+  typeof data === "object" && data !== null && "ii_credential_response" in data;
+
+const requestCredential = (
+  options: CredentialRequestOptions,
+  targetWindow: Window,
+  targetOrigin: string
+): Promise<PublicKeyCredential> =>
+  new Promise<PublicKeyCredential>((resolve, reject) => {
+    // Listen for credential response
+    const id = window.crypto.randomUUID();
+    const listener = (event: MessageEvent) => {
+      if (
+        event.source !== targetWindow ||
+        event.origin !== targetOrigin ||
+        !isCredentialResponse(event.data)
+      ) {
+        return;
+      }
+      if ("result" in event.data.ii_credential_response) {
+        resolve({
+          ...event.data.ii_credential_response.result,
+          getClientExtensionResults: () => ({}),
+        } as PublicKeyCredential);
+      }
+      if ("err" in event.data.ii_credential_response) {
+        reject(event.data.ii_credential_response.err);
+      }
+      window.removeEventListener("message", listener);
+    };
+    window.addEventListener("message", listener);
+
+    // Request credential
+    targetWindow.postMessage(
+      {
+        ii_credential_request: { id, options },
+      } as CredentialRequest,
+      targetOrigin
+    );
+  });
+
+const handleCredentialRequest = (
+  targetWindow: Window,
+  targetOrigin: string
+): void =>
+  window.addEventListener("message", async (event: MessageEvent) => {
+    if (
+      event.source === targetWindow &&
+      event.origin === targetOrigin &&
+      isCredentialRequest(event.data)
+    ) {
+      try {
+        const credential = (await navigator.credentials.get(
+          event.data.ii_credential_request.options
+        )) as PublicKeyCredential;
+        window.parent.postMessage(
```

Comment:
Could you create the object result object before and type it with the expected type?

It's hard to read otherwise.